src/libs/Violations/useViolations.ts

```diff
@@ -0,0 +1,67 @@
+import {useCallback, useMemo} from 'react';
+import useLocalize from '@hooks/useLocalize';
+import {TransactionViolation, ViolationName} from '@src/types/onyx';
+
+/**
+ * Map from Violation Names to the field where that violation can occur
+ */
+const violationFields: Record<ViolationName, ViolationField> = {
+    perDayLimit: 'amount',
+    maxAge: 'date',
+    overLimit: 'amount',
+    overLimitAttendee: 'amount',
+    overCategoryLimit: 'amount',
+    receiptRequired: 'receipt',
+    missingCategory: 'category',
+    categoryOutOfPolicy: 'category',
+    missingTag: 'tag',
+    tagOutOfPolicy: 'tag',
+    missingComment: 'comment',
+    taxRequired: 'tax',
+    taxOutOfPolicy: 'tax',
+    billableExpense: 'billable',
+};
+
+/**
+ * Names of Fields where violations can occur
+ */
+type ViolationField = 'merchant' | 'amount' | 'category' | 'date' | 'tag' | 'comment' | 'billable' | 'receipt' | 'tax';
+
+type ViolationsMap = Map<ViolationField, TransactionViolation[]>;
+
+export default function useViolations(violations: TransactionViolation[]) {
+    const {translate} = useLocalize();
+
+    const violationsByField = useMemo((): ViolationsMap => {
+        const violationGroups = new Map<ViolationField, TransactionViolation[]>();
```

Comment:
My hot take on hooks is that all plain functions within them should be extracted and then consumed within the hook, for two reasons:

1. Easier testing (less relevant here)
2. Easier re-use of code (there are places where hooks can't be used easily but the logic might still be relevant).